streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreBuilderWrapper.java

```diff
@@ -0,0 +1,119 @@
+import org.apache.kafka.streams.state.internals.VersionedKeyValueStoreBuilder;
+import org.apache.kafka.streams.state.internals.WindowStoreBuilder;
+
+public class StoreBuilderWrapper<S extends StateStore> implements StoreFactory<S> {
+
+    private final StoreBuilder<S> builder;
+    private final Set<String> users = new HashSet<>();
+
+    public StoreBuilderWrapper(final StoreBuilder<S> builder) {
+        this.builder = builder;
+    }
+
+    @Override
+    public S build() {
+        return builder.build();
+    }
+
+    @Override
+    public long retentionPeriod() {
+        if (builder instanceof WindowStoreBuilder) {
+            return ((WindowStoreBuilder<?, ?>) builder).retentionPeriod();
+        } else if (builder instanceof TimestampedWindowStoreBuilder) {
+            return ((TimestampedWindowStoreBuilder<?, ?>) builder).retentionPeriod();
+        } else if (builder instanceof SessionStoreBuilder) {
+            return ((SessionStoreBuilder<?, ?>) builder).retentionPeriod();
+        } else {
+            throw new IllegalStateException(
+                    "retentionPeriod is not supported when not a window store");
+        }
+    }
+
+    @Override
+    public long historyRetention() {
+        if (builder instanceof VersionedKeyValueStoreBuilder) {
+            return ((VersionedKeyValueStoreBuilder<?, ?>) builder).historyRetention();
+        } else {
+            throw new IllegalStateException(
+                    "historyRetention is not supported when not a versioned store");
+        }
+    }
+
+    @Override
+    public Set<String> users() {
+        return users;
+    }
+
+    @Override
+    public boolean loggingEnabled() {
+        return builder.loggingEnabled();
+    }
+
+    @Override
+    public String name() {
+        return builder.name();
+    }
+
+    @Override
+    public boolean isWindowStore() {
+        return builder instanceof WindowStoreBuilder
+                || builder instanceof TimestampedWindowStoreBuilder
+                || builder instanceof SessionStoreBuilder;
+    }
+
+    @Override
+    public boolean isVersionedStore() {
+        return builder instanceof VersionedKeyValueStoreBuilder;
+    }
+
+    // Apparently Java strips the generics from this method because we're using the raw type for builder,
+    // even though this method doesn't use builder's (missing) type parameter. Our usage seems obviously
+    // correct, though, hence the suppression.
+    @Override
+    public Map<String, String> logConfig() {
+        return builder.logConfig();
+    }
+
+    @Override
+    public StoreFactory<S> withCachingDisabled() {
+        builder.withCachingDisabled();
+        return this;
+    }
+
+    @Override
+    public StoreFactory<S> withLoggingDisabled() {
+        builder.withLoggingDisabled();
+        return this;
+    }
+
+    @Override
+    public boolean isCompatibleWith(final StoreFactory<?> storeFactory) {
```

Comment:
I know you didn't add this code (even if the method itself is new) but some javadocs or a more descriptive name might be useful here. Hard to understand whether this is/will be working correctly after the changes when it's unclear what its original semantics/purpose were

I guess this is related to a fundamental source of confusion I have right now, which is: where does the StoreBuilderWrapper fit into the final plan? Will anything extend it, or will uninitialized/unconfigured StoreFactory implementations somehow be compiled into this class when they finally set up a real StoreBuilder? Like is this class a "storebuilder-wrapper" in the sense that it always wraps an actual configured/initialized StoreBuilder object, or will the `builder` field of this class be "delayed" (in which case how do we evaluate if it's compatible, hence how it all relates to this specific question/method)

Reply:
I'll add a comment. The plan is to add a separate `DelayedStoreFactory` which delays the creation of the `StoreBuilder` until `configure` is called. the `isCompatibleWith` method will check to make sure the inner components would generate the same store builder (e.g. the wrapped `StoreSuppliers` are identical)

I'll just make that change and add it to this PR to show how it works end-to-end, but that will be a functionality change.